Simplenote/ExcerptMaker.swift

```diff
@@ -0,0 +1,61 @@
+import Foundation
+
+// MARK: - ExcerptMaker: Generate excerpt from a note with specified keywords
+//
+@objc
+final class ExcerptMaker: NSObject {
+    private var regexp: NSRegularExpression?
+    private var keywords: [String]?
+
+    /// Update with search keywords
+    ///
+    @objc
+    func update(withKeywords keywords: [String]?) {
+        guard keywords != self.keywords else {
+            return
+        }
+
+        self.keywords = keywords
+
+        guard let keywords = keywords, !keywords.isEmpty else {
+            regexp = nil
+            return
+        }
+
+        regexp = NSRegularExpression.regexForExcerpt(withKeywords: keywords)
+    }
+
+    /// Generate and return excerpt from the note. Excerpt is based on keywords
+    ///
+    func excerpt(from note: Note) -> String? {
+        guard let regexp = regexp, let content = note.content else {
+            return note.bodyPreview
+        }
+
+        let foldedContent = content.folding(options: [.diacriticInsensitive], locale: .current)
+
+        let bodyRange = NoteContentHelper.structure(of: foldedContent).body
+        guard bodyRange.location != NSNotFound else {
+            return nil
+        }
+
+        let excerptRange = regexp.rangeOfFirstMatch(in: foldedContent,
+                                                    options: [],
+                                                    range: bodyRange)
```

Comment:
Idea: how about if, instead of using `NoteContentHelper` to calculate the bodyRange we just do...

```
let bodyText = note.bodyPreview
let excerptRange = regexp.rangeOfFirstMatch(in: bodyText, options: [], range: bodyText.fullRange)
```